dbrp/service.go

```diff
@@ -0,0 +1,489 @@
+func (s *Service) getDefaultID(ctx context.Context, orgID influxdb.ID, db string) (influxdb.ID, error) {
+	defID, err := s.getDefault(ctx, orgID, db)
+	if err != nil {
+		return 0, err
+	}
+	id := new(influxdb.ID)
+	if err := id.Decode(defID); err != nil {
+		return 0, err
+	}
+	return *id, nil
+}
+
+// isDefault tells whether a mapping is the default one.
+func (s *Service) isDefault(ctx context.Context, m *influxdb.DBRPMappingV2) (bool, error) {
+	defID, err := s.getDefault(ctx, m.OrganizationID, m.Database)
+	if err != nil {
+		if kv.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	dbrpID, _ := m.ID.Encode()
+	return bytes.Equal(dbrpID, defID), nil
+}
+
+// updateDefault updates the default value of the given mapping based on the default value stored.
+func (s *Service) updateDefault(ctx context.Context, m *influxdb.DBRPMappingV2) error {
+	if d, err := s.isDefault(ctx, m); err != nil {
+		return err
+	} else if d {
+		m.Default = true
+	} else {
+		m.Default = false
+	}
+	return nil
+}
+
+// isDBRPUnique verifies if the triple orgID-database-retention-policy is unique.
+func (s *Service) isDBRPUnique(ctx context.Context, m influxdb.DBRPMappingV2) error {
+	return s.store.View(ctx, func(tx kv.Tx) error {
+		return s.byOrgAndDatabase.Walk(ctx, tx, composeForeignKey(m.OrganizationID, m.Database), func(k, v []byte) error {
+			dbrp := &influxdb.DBRPMappingV2{}
+			if err := json.Unmarshal(v, dbrp); err != nil {
+				return ErrInternalService(err)
+			}
+			if dbrp.ID == m.ID {
+				// Corner case.
+				// This is the very same DBRP, just skip it!
+				return nil
+			}
+			if dbrp.RetentionPolicy == m.RetentionPolicy {
+				return ErrDBRPAlreadyExists("another DBRP mapping with same orgID, db, and rp exists")
+			}
+			return nil
+		})
+	})
+}
+
+// getDefaultTx gets the default mapping ID inside of a transaction.
+func (s *Service) getDefaultTx(tx kv.Tx, compKey []byte) ([]byte, error) {
+	b, err := tx.Bucket(defaultBucket)
+	if err != nil {
+		return nil, err
+	}
+	defID, err := b.Get(compKey)
+	if err != nil {
+		return nil, err
+	}
+	return defID, nil
+}
+
+// setDefaultTx sets the given ID as default according to the given default value.
+// setDefaultTx ensures that the first value for the given composite key is always the default,
+// disrespectfully of the given default value.
+// NOTE: updates to the bucket happen in a transaction.
+func (s *Service) setDefaultTx(tx kv.Tx, compKey []byte, id []byte, defValue bool) (bool, error) {
+	set := defValue
+	b, err := tx.Bucket(defaultBucket)
+	if err != nil {
+		return defValue, ErrInternalService(err)
+	}
+	_, err = b.Get(compKey)
+	if err != nil {
+		if kv.IsNotFound(err) {
+			// This is the only one.
+			// This has to be default, no matter defValue.
+			set = true
+		} else {
+			return defValue, ErrInternalService(err)
+		}
+	}
+	if set {
+		if err := b.Put(compKey, id); err != nil {
+			return defValue, ErrInternalService(err)
+		}
+	}
```

Comment:
I tried to walk through this behaviour and see if I could make it more readable. I came to this which I think is equivalent, but I could be mistaken:
```go
	_, err = b.Get(compKey)
	// given a nil error (dbrp is set as default) and we do not want to set the dbrp as the default
	if err == nil && !defValue  {
		// then just return default value as provided
		return defValue, nil
	// given a nil error (dbrp is set as default) and we want to set the dbrp as the default
	// or the key was not present (therefore the dbrp is not currently the default)
	} else if (err == nil && defValue) || kv.IsNotFound(err) {
		if err = b.Put(compKey, id); err == nil {
			return true, nil
		}
	}
	return defValue, ErrInternalService(err)
```

Which I think is equivalent and when I read it back it makes me wonder about the behaviour of the method.

There are couple cases that make me double take:
1. if `defValue == false (we want to unset) && the DBRP is currently set as default` it will noop and return `false`.
2. given `the DBRP is currently set as default && defValue == true` it will set it again.

Taking this method out of the context where it is called, these behaviours confused me. Perhaps they make sens in the called context. So I will dig further into where they're called and perhaps this will clear up my understanding :D

Reply:
I simplified the methods in the last commit. Everything should be clearer.